### PR TITLE
Add range request functionality

### PR DESCRIPTION
**Pull request by dshirley**
A basic implementation of range requests:
http://svn.tools.ietf.org/svn/wg/httpbis/specs/rfc7233.html

Allows querying http://httpbin.org/range-request/1024?duration=s&chunk_size=code, specifying Range headers that allow selecting specific portions of the resource to GET.

### CODE REVIEW

File: httpbin/core.py [hunk 1]
```diff
@@ -537,10 +537,91 @@ def generate_bytes():
             yield(bytes(chunks))
 
     headers = {'Transfer-Encoding': 'chunked',
-               'Content-Type': 'application/octet-stream'}
+               'Content-Type': 'application/octet-stream' }
```
Comment (sigmavirus24) [inline httpbin/core.py]: Why did you change this? It's not a style improvement and it has no other benefits. Please remove the trailing space.

Comment (dshirley) [inline httpbin/core.py]: ah, I think this was an accident.

**Comment by sigmavirus24** [comment]
Why does this need to be a separate endpoint? You're duplicating portions of other endpoints (it seems). Why not abstract that code into a separate module and use those common portions instead of copying code around?

**Comment by dshirley** [comment]
Yep, you're correct.  This endpoint is an amalgamation of code in /stream-bytes and /drip.
I would much prefer that this functionality IS a part of one (or both) of these endpoints.  However, I didn't do that because it requires changing the behavior of those endpoints.  The behavior changes might be negligible enough to where they don't matter - I'll let you be the judge.

Here's the requirements that didn't jive with the other end points.

1)  The payload generation has to be deterministic.  /stream-bytes generates random data.  /stream-bytes even allows for a random seed.  The seed parameter would have to be removed.
2)  I need chunked transfer encoding.  This doesn't seem to be in the RFC, but one of the things I'm hoping to test is NSURLSession downloadTaskWithResumeData (https://developer.apple.com/library/ios/documentation/Foundation/Reference/NSURLSession_class/#//apple_ref/occ/instm/NSURLSession/downloadTaskWithResumeData:).  It seems that Apple requires the transfer-encoding be "chunked" in order for this to work.  The bottom line is "/drip" currently uses "identity" encoding rather than chunked.
3) The data has to be a predictable length.  /stream-bytes takes the ceiling of the input number of bytes with 100k.  It would be pretty wonky to request bytes 100k-200k and also have the endpoint automatically clip the amount of data you requested.

There is one other benefit and that is having a separate endpoint makes it a bit more obvious which endpoint supports range-requests - much like the "stream-bytes" endpoint makes it obvious which endpoint supports chunked transfer-encoding.

Are you guys ok with modifying the behavior of any of the other endpoints?  I'm also fine factoring out the generating function, though it is fairly different than the other generating functions, so I'm not sure how much benefit will come of it as far as code deduplication goes (though it maybe more readable).

Thoughts?

**Comment by johnsheehan** [comment]
i'll let you all sort out the refactoring issues...my only question here is the path name. is `-request` required or would just `/range` suffice?

**Comment by dshirley** [comment]
@johnsheehan I changed the endpoint from /range-request to /range

**Comment by dshirley** [comment]
@sigmavirus24 Do you have any particular suggestions on what should be factored out?  The only piece that is really copy/pasted is the part that gets the chunk_size from the URL parameters:

      if 'chunk_size' in params:
          chunk_size = max(1, int(params['chunk_size']))
      else:
          chunk_size = 10 * 1024

There are a few other places where I borrowed inspiration, but the logic seems different enough that refactoring the code would require some unnatural gymnastics that wouldn't necessarily result in more readable code.

I did factor out a piece of logic to make the code a bit more concise.  I'm open to further changes.  I just want to make sure we're in sync with the direction we want it to go before I spend time doing the work.

**Comment by dshirley** [comment]
@johnsheehan Can this be merged?

**Comment by johnsheehan** [comment]
@sigmavirus24 and others, any last comments on the code? i'm cool with the functionality and route name.

**Comment by sigmavirus24** [comment]
@johnsheehan one request and then I'm okay with this.

**Comment by dshirley** [comment]
@johnsheehan @sigmavirus24 I removed the changes to the .gitignore by doing a "git reset HEAD~1; git push -f".

**Comment by sigmavirus24** [comment]
Last I checked I was :+1: on the code

**Comment by johnsheehan** [comment]
will deploy to httpbin.org in the next week or two (swamped with other projects now)

**Comment by dshirley** [comment]
Good working with you guys.  Much appreciated!
